pdns/opensslsigners.cc

```diff
@@ -274,7 +332,24 @@ void OpenSSLRSADNSCryptoKeyEngine::create(unsigned int bits)
     throw runtime_error(getName() + " key generation failed while setting e");
   }
 
-  auto key = std::unique_ptr<RSA, decltype(&RSA_free)>(RSA_new(), RSA_free);
+#if OPENSSL_VERSION_MAJOR >= 3
+  auto ctx = KeyContext(EVP_PKEY_CTX_new_from_name(nullptr, "RSA", nullptr), EVP_PKEY_CTX_free);
+  if (ctx == nullptr) {
+    throw pdns::OpenSSL::error(getName(), "Could not initialize context");
+  }
+
+  if (EVP_PKEY_keygen_init(ctx.get()) != 1) {
+    throw pdns::OpenSSL::error(getName(), "Could not initialize keygen");
+  }
+
+  EVP_PKEY* key = nullptr;
```

Comment:
Don't we need to call `EVP_PKEY_CTX_set_rsa_keygen_bits(ctx.get(), bits)` and `EVP_PKEY_CTX_set1_rsa_keygen_pubexp(ctx.get(), exponent.get())` first? The documentation states that the default exponent is 65537 anyway, but it seems safer to set it explicitly, especially since we already computed it above.